gr-zeromq/lib/rep_msg_sink_impl.cc

```diff
@@ -79,7 +84,7 @@ void rep_msg_sink_impl::readloop()
     while (!d_finished) {
 
         // while we have data, wait for query...
-        while (!empty_p(d_port)) {
+        while (!empty_p(d_port) && !d_finished) {
```

Comment:
Why is checking `d_finished` necessary here? Some lines below there is
```c++
               if (!ok) {
                    // Should not happen, we've checked POLLIN.
                    d_logger->error("Failed to receive message.");
                    break; // Fall back to re-check d_finished
                }
```
This `break` goes to the `d_finished` check in the outermost `while` loop. Are there other cases in which `d_finished` can become `true` that are not handled by this code?

I find the logic in `readloop()` slightly difficult to follow, so this is why I wonder what are exactly the changes achieved by introducing another `d_finished` check here.

Reply:
It looks like it bypasses the poll, terminating d_timeout ms faster.

Reply:
The other reason is 
- if the producer of the messages is sending them faster than block can pull them, there's a chance that the socket/context have been closed before we check if we're finished.
- This condition will cause errors on exit that having the check if finished there will naturally handle.

Reply:
> It looks like it bypasses the poll, terminating d_timeout ms faster.

Okay. That makes sense on a second read.

> there's a chance that the socket/context have been closed before we check if we're finished

How? We only close the socket and context when we destroy the class. This thread is joined in `stop()`. How can the destructor run before the thread finishes?

Reply:
Sorry I'm confusing problems.

The problem is that the call to close can hang under the conditions of:
- d_port has more messages
- stop() has been called
- the state of the block is waiting for POLLIN to be set, but the sender is already done.

Under these conditions the inner loop is infinite. Checking d_finished there prevents that's.

Reply:
That explanation makes total sense. Thanks. Consider this solved (I think I can't mark the thread as solved myself).